File: backend/src/main/java/com/productivity/service/UserService.java
package com.productivity.service;

import com.google.cloud.Timestamp;
import com.productivity.dto.UserDTO;
import com.productivity.exception.ResourceNotFoundException;
import com.productivity.model.User;
import com.productivity.repository.UserRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ExecutionException;

@Service
public class UserService {

    private static final Logger log = LoggerFactory.getLogger(UserService.class);
    private final UserRepository userRepository;

    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public UserDTO createOrUpdateUser(String uid, String email, String displayName, String photoURL)
            throws ExecutionException, InterruptedException {
        String normalizedEmail = email != null ? email.toLowerCase().trim() : null;
        User user = userRepository.findById(uid).orElse(null);

        if (user == null) {
            user = new User();
            user.setUid(uid);
            user.setEmail(normalizedEmail);
            user.setDisplayName(displayName);
            user.setPhotoURL(photoURL);
            user.setRole("user");
            user.setOnboarded(false);
            user.setVerified(false); // Default to false for new enterprise auth
            user.setCreatedAt(Timestamp.now());
            user.setUpdatedAt(Timestamp.now());
            userRepository.save(user);
            log.info("New user created: {}", uid);
        } else {
            Map<String, Object> updates = new HashMap<>();
            if (normalizedEmail != null) updates.put("email", normalizedEmail);
            if (displayName != null) updates.put("displayName", displayName);
            if (photoURL != null) updates.put("photoURL", photoURL);
            updates.put("updatedAt", Timestamp.now());
            userRepository.update(uid, updates);
            user = userRepository.findById(uid).orElseThrow();
            log.info("User updated: {}", uid);
        }

        return toDTO(user);
    }
    
    // New method for enterprise signup
    public User createUser(String email, String passwordHash, String displayName) 
            throws ExecutionException, InterruptedException {
        String normalizedEmail = email != null ? email.toLowerCase().trim() : null;
        String uid = UUID.randomUUID().toString();
        
        User user = new User();
        user.setUid(uid);
        user.setEmail(normalizedEmail);
        user.setPasswordHash(passwordHash);
        user.setDisplayName(displayName);
        user.setRole("user");
        user.setOnboarded(false);
        user.setVerified(false);
        user.setCreatedAt(Timestamp.now());
        user.setUpdatedAt(Timestamp.now());
        
        userRepository.save(user);
        log.info("Enterprise user created: {}", uid);
        return user;
    }
    
    public User findByEmail(String email) throws ExecutionException, InterruptedException {
        if (email == null) return null;
        String normalizedEmail = email.toLowerCase().trim();
        return userRepository.findByEmail(normalizedEmail).orElse(null);
    }

    public UserDTO getUser(String uid) throws ExecutionException, InterruptedException {
        User user = getUserModel(uid);
        return toDTO(user);
    }

    public User getUserModel(String uid) throws ExecutionException, InterruptedException {
        return userRepository.findById(uid)
                .orElseThrow(() -> new ResourceNotFoundException("User", uid));
    }

    public UserDTO completeOnboarding(String uid, String bio) throws ExecutionException, InterruptedException {
        Map<String, Object> updates = new HashMap<>();
        updates.put("bio", bio);
        updates.put("onboarded", true);
        updates.put("updatedAt", Timestamp.now());
        
        User user = userRepository.update(uid, updates);
        log.info("Onboarding completed for user: {}", uid);
        return toDTO(user);
    }
    
    public void verifyUser(String uid) throws ExecutionException, InterruptedException {
         Map<String, Object> updates = new HashMap<>();
         updates.put("verified", true);
         updates.put("updatedAt", Timestamp.now());
         userRepository.update(uid, updates);
    }
    
    public void recordLogin(String uid) throws ExecutionException, InterruptedException {
        Map<String, Object> updates = new HashMap<>();
        updates.put("lastLogin", Timestamp.now());
        updates.put("loginAttempts", 0);
        updates.put("lockedUntil", null);
        userRepository.update(uid, updates);
    }
    
    public void updatePassword(String uid, String newPasswordHash) throws ExecutionException, InterruptedException {
        Map<String, Object> updates = new HashMap<>();
        updates.put("passwordHash", newPasswordHash);
        updates.put("updatedAt", Timestamp.now());
        userRepository.update(uid, updates);
    }

    public UserDTO toDTO(User user) {
        UserDTO dto = new UserDTO();
        dto.setUid(user.getUid());
        dto.setEmail(user.getEmail());
        dto.setDisplayName(user.getDisplayName());
        dto.setPhotoURL(user.getPhotoURL());
        dto.setRole(user.getRole());
        dto.setOnboarded(user.isOnboarded());
        dto.setBio(user.getBio());
        
        dto.setVerified(user.isVerified());
        dto.setLoginAttempts(user.getLoginAttempts());
        if (user.getLockedUntil() != null) {
            dto.setLockedUntil(user.getLockedUntil().toDate().toInstant().toString());
        }
        if (user.getLastLogin() != null) {
            dto.setLastLogin(user.getLastLogin().toDate().toInstant().toString());
        }
        
        if (user.getCreatedAt() != null) {
            dto.setCreatedAt(user.getCreatedAt().toDate().toInstant().toString());
        }
        return dto;
    }
}

File: backend-node/src/services/userService.js
'use strict';

const { v4: uuidv4 } = require('uuid');
const userRepo = require('../repositories/userRepository');
const { sendSimpleMessage } = require('../utils/email');
const { toISO } = require('../utils/timestamp');

async function createOrUpdateGoogleUser(uid, email, displayName, photoURL) {
  const normalizedEmail = email ? email.toLowerCase().trim() : null;
  let user = await userRepo.findById(uid);

  if (!user) {
    user = {
      uid,
      email: normalizedEmail,
      displayName: displayName || null,
      photoURL: photoURL || null,
      role: 'user',
      onboarded: false,
      verified: false,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    };
    await userRepo.save(user);
  } else {
    const updates = { updatedAt: new Date().toISOString() };
    if (normalizedEmail) updates.email = normalizedEmail;
    if (displayName) updates.displayName = displayName;
    if (photoURL) updates.photoURL = photoURL;
    user = await userRepo.update(uid, updates);
  }

  return toDTO(user);
}

async function createUser(email, passwordHash, displayName) {
  const normalizedEmail = email ? email.toLowerCase().trim() : null;
  const uid = uuidv4();

  const user = {
    uid,
    email: normalizedEmail,
    passwordHash,
    displayName: displayName || null,
    photoURL: null,
    role: 'user',
    onboarded: false,
    verified: false,
    loginAttempts: 0,
    lockedUntil: null,
    lastLogin: null,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString(),
  };

  await userRepo.save(user);
  return user;
}

async function findByEmail(email) {
  if (!email) return null;
  return userRepo.findByEmail(email.toLowerCase().trim());
}

async function getUser(uid) {
  const user = await userRepo.findById(uid);
  if (!user) throw Object.assign(new Error(`User not found: ${uid}`), { status: 404 });
  return toDTO(user);
}

async function getUserModel(uid) {
  const user = await userRepo.findById(uid);
  if (!user) throw Object.assign(new Error(`User not found: ${uid}`), { status: 404 });
  return user;
}

async function verifyUser(uid) {
  await userRepo.update(uid, { verified: true, updatedAt: new Date().toISOString() });
}

async function recordLogin(uid) {
  await userRepo.update(uid, {
    lastLogin: new Date().toISOString(),
    loginAttempts: 0,
    lockedUntil: null,
    updatedAt: new Date().toISOString(),
  });
}

async function incrementLoginAttempts(uid, attempts) {
  await userRepo.update(uid, {
    loginAttempts: attempts,
    updatedAt: new Date().toISOString(),
  });
}

async function lockAccount(uid, lockedUntil) {
  await userRepo.update(uid, {
    loginAttempts: 5,
    lockedUntil,
    updatedAt: new Date().toISOString(),
  });
}

async function updatePassword(uid, newPasswordHash) {
  await userRepo.update(uid, { passwordHash: newPasswordHash, updatedAt: new Date().toISOString() });
}

async function completeOnboarding(uid, bio) {
  const user = await userRepo.update(uid, {
    bio,
    onboarded: true,
    updatedAt: new Date().toISOString(),
  });
  return toDTO(user);
}

async function updateProfile(uid, displayName, bio, photoURL) {
  const updates = { updatedAt: new Date().toISOString() };
  if (displayName != null) updates.displayName = displayName;
  if (bio != null) updates.bio = bio;
  if (photoURL != null) updates.photoURL = photoURL;
  const user = await userRepo.update(uid, updates);
  return toDTO(user);
}

async function deleteUser(uid) {
  const user = await userRepo.findById(uid);
  const email = user?.email;
  const displayName = user?.displayName;

  await userRepo.remove(uid);

  if (email) {
    sendSimpleMessage(
      email,
      'Your Productiv Account Has Been Deleted',
      `Hi ${displayName || 'there'},<br><br>Your Productiv account has been permanently deleted and all associated data removed.<br><br>If you did not request this, please contact support immediately.<br><br>The Productiv Team`
    ).catch(() => {});
  }
}

function toDTO(user) {
  return {
    uid: user.uid,
    email: user.email,
    displayName: user.displayName,
    photoURL: user.photoURL,
    role: user.role,
    onboarded: user.onboarded,
    bio: user.bio || null,
    verified: user.verified,
    loginAttempts: user.loginAttempts || 0,
    lockedUntil: toISO(user.lockedUntil),
    lastLogin: toISO(user.lastLogin),
    createdAt: toISO(user.createdAt),
    xp: user.xp || 0,
    level: user.level || 1,
    totalTasksCompleted: user.totalTasksCompleted || 0,
    streakBest: user.streakBest || 0,
  };
}

module.exports = {
  createOrUpdateGoogleUser,
  createUser,
  findByEmail,
  getUser,
  getUserModel,
  verifyUser,
  recordLogin,
  incrementLoginAttempts,
  lockAccount,
  updatePassword,
  completeOnboarding,
  updateProfile,
  deleteUser,
  toDTO,
};
